### DISCUSSION TITLE
Spacy for invoice data extraction

### DISCUSSION THREAD

**Discussion by ssherlins** [explosion/spaCy#8187] [Help: Other Questions] [answered]
Hi,
I recently came across Spacy when searching for a ML model to extract data from invoices. We have many different formats (300+) of invoices, so I was looking to see if there is a way to train an ML model which can extract details like invoice number, purchase order number, invoice date etc. So that's when I read about NER and Spacy. The invoices are in PDF format, most of which are scanned. So I used Tesseract to extract data from PDF into text files and used that data to annotate and train. The extracted data had a lot of empty lines. Will that affect the training? I used the following annotator: https://github.com/tecoholic/ner-annotator. I also want to know whether it is right to use NER for data extraction from invoices with multiple formats.

**Comment by polm** [comment, marked as answer]
You should certainly be able to train an NER model on your data. I guess I'd need to see some example invoices to say if spaCy would have trouble with them or not, but it should be worth a shot at least. 

A lot of empty lines is not important. You can just remove them. 

One issue would be whether tesseract had done a good job. If your text is mangled spaCy may not be able to recover anything. This would depend on how good the tesseract model is, how clean your images are, and other details. 

Another issue is whether layout or word sequences are important. For example, if your invoice number is always in the top right of the page, spaCy has no way of knowing that. But if your document looks like this:

> Invoice number: 1235456

Then spaCy can probably learn that, or you can use the rule based matchers to find it. spaCy is good at learning from sequences of words, so if you just sort of have a random list of numbers and a few words it may be hard to learn.

If you can provide some examples we can probably give more tailored advice.

**Reply by ssherlins** [reply]
Thank you. Most of my invoices are scanned, so the extracted data is not very good. I used tesseract for extraction. 
I'm sorry, the invoices are confidential, so I'll not be able to post them here.

**Reply by polm** [reply]
OK, in that case it's hard to say much. Maybe you can generate a test invoice you can share? But if the data is very messy it might be hard to use with normal methods in spaCy or other libraries. You might try starting with the rule-based matchers and seeing how well that works as a baseline. 

https://spacy.io/usage/rule-based-matching

**Reply by ssherlins** [reply]
[PDF1111111.txt](https://github.com/explosion/spaCy/files/6538442/PDF1111111.txt)
Hi, I have uploaded the result of OCR on one of our invoices. I am using the following code to train the model:

```
import json
import spacy

# load the training data
with open('training_data.json') as fp:
  training_data = json.load(fp)


# prepare an empty model to train
nlp = spacy.blank('en')
nlp.vocab.vectors.name = 'demo'
ner = nlp.create_pipe('ner')
nlp.add_pipe(ner, last=True)

# Add the custome NER Tags as entities into the model
for label in training_data["classes"]:
    print("label",label)
    nlp.entity.add_label(label)

# Train the model
optimizer = nlp.begin_training()
for iter in range(50):
    print(iter)
    for text, annotations in training_data["annotations"]:
        if len(text) > 0: # in case an empty sentence was saved while annotating
            
            nlp.update([text], [annotations], sgd=optimizer)
```

My json file after annotation is as follows. 
[annotated.txt](https://github.com/explosion/spaCy/files/6538679/annotated.txt)



I am getting an empty list when I use the same data to test. 

Also this is just one invoice. How do I train for multiple invoices?

**Reply by walidamamou** [reply]
Hi,

Having document layout information for model training is crucial to get high scores (it is not an option). You need to provide entity labels along with their coordinates. I recommend to checkout layoutlm from microsoft: https://github.com/microsoft/unilm/tree/master/layoutlm

**Reply by ssherlins** [reply]
Hi @walidamamou, I didn't understand it. I am a beginner in this. Could you please tell me why layoutlm should be used and how ? Also how does Spacy take the data as input for training. If you could provide some docs/links, it would be helpful.

Thank you.

**Reply by ssherlins** [reply]
Hi, I would also like to understand whether the whole document should be included in the annotated json file or only the lines which contain the annotation? And does the json/txt file having utf-8 encoding have an effect on the training and testing? 

Thank you

**Reply by walidamamou** [reply]
No problem. Here is the original  article describing how it works: https://arxiv.org/pdf/1912.13318.pdf
But in summary the LayoutLM model uses transformer architecture in addition to adding two type of inputs:
- a 2-D position embedding that denotes the relative position of a token within a document
-  an image embedding for scanned token images within a document

Because text in invoices are sparse and there is no much semantic context around the tokens, regular NER model won't perform well on this type of task. I doubt the spacy model will perform well since it doesn't take in account the 2D spatial position of each token. The position of tokens contain valuable information for this type of extraction and needs to be included.

The layouLM model is available in [huggingface](https://huggingface.co/transformers/model_doc/layoutlm.html) and can be fine tuned as shown [here](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForTokenClassification_on_FUNSD.ipynb#scrollTo=DTFnEZagQm4v)

**Reply by polm** [reply]
Hello @walidamamou, welcome to the thread. 

@ssherlins To be clear, I think @walidamamou is recommending a non-spaCy approach. spaCy doesn't have a good way to use data not present directly in text data, like information about the position on the page or image data. 

If your OCR data is messy, or your page layout is important to understanding the invoice content, that data is very important. It's very common for that to be the case.

However, given your example document, you literally have a line that's `Invoice Number xxxxxxx`. You don't need a fancy model to extract that. I would recommend you carefully read the [rule based matching docs](https://spacy.io/usage/rule-based-matching), spend a little time implementing a rule-based model, and see how far it gets you. Even if you don't end up using spaCy's NER that would make a good baseline to compare against any future system you make.

spaCy's pretrained models will also probably be able to identify dates in your document.

Now, regarding your NER model, you can't train successfully with just one document. I'd suggest you have at least 100, preferably 300 before getting started. 

I suggest you use the training config rather than a custom training loop. You can also check the [example projects](https://github.com/explosion/projects/tree/v3/tutorials) for detailed examples of creating training data, but basically you just create Docs that look like you want your output to look and then save them into a DocBin. 

Also, it's a few years old, but [this thread about invoice NER](https://support.prodi.gy/t/invoice-parsing-using-spacy/728) from the Prodigy forums may be helpful.

Hope that helps, let me know if you have any other questions.

**Reply by ssherlins** [reply]
Thank you @walidamamou, I'll check it out.

@polm thank you for the clarification. 

The invoices that I have do not conform to a specific format (for example, "Invoice number" may be mentioned as "Document Number" or simply "No."). So I'm not sure whether rule based matching docs will be useful in this case. I had tried Google Cloud Platform's (GCP) AutoML using NER and that's how I learnt about NER. It served the purpose quite well, unfortunately GCP's AutoML does not provide an option to retrain an existing model, and that's how I started searching for NER and came across spaCy. I think I have to continue researching on both spaCy and LayoutLM. 

Can you please confirm how I can train data from multiple invoices? should I append the data and create a single json file or run the training on each json file?
Also. should the json file contain the lines in which I haven't annotated any entity (for example, if I want to create an NER model to extract invoice number and date, and my invoice doc contains multiple lines like company name, address etc. should I include them in the json file or only the data that I want i.e. invoice-no)?

Thank you

**Reply by polm** [reply]
> The invoices that I have do not conform to a specific format (for example, "Invoice number" may be mentioned as "Document Number" or simply "No."). So I'm not sure whether rule based matching docs will be useful in this case. 

Got it, that makes sense. The nice thing about rule-based matching is that even if you don't end up using it, it should be *very* fast to implement, and when you try other methods you can compare them to it.

> I had tried Google Cloud Platform's (GCP) AutoML using NER and that's how I learnt about NER. It served the purpose quite well

I have not used Google's AutoML, but it just works on text, right? If that's the case and you found it effective then that's actually a good sign your data doesn't require spatial information.

> Can you please confirm how I can train data from multiple invoices? should I append the data and create a single json file or run the training on each json file?

You don't have to make a json file to train a model, whether it's json or not is kind of unrelated. Instead what you do is make the Docs you want manually and then save them in a DocBin. If you look at [this example script](https://github.com/explosion/projects/blob/v3/tutorials/ner_fashion_brands/scripts/preprocess.py) you'll see that it creates Docs and assigns entities in a loop. That's what you should do. 

The DocBin file will then have many documents in it, and you can specify it as training data when using the config.

> Also. should the json file contain the lines in which I haven't annotated any entity 

Yes, it's very important it has examples with no entity. If all your training data has entities, then your model will learn to always mark an entities, but not all your data will have invoice numbers or whatever. Your training data should be as close to real data as possible.

**Reply by ssherlins** [reply]
> You don't have to make a json file to train a model, whether it's json or not is kind of unrelated. Instead what you do is make the Docs you want manually and then save them in a DocBin. If you look at [this example script](https://github.com/explosion/projects/blob/v3/tutorials/ner_fashion_brands/scripts/preprocess.py) you'll see that it creates Docs and assigns entities in a loop. That's what you should do.

I went through the script. I still don't understand the format of the training data. Is Doc the image of the invoice and DocBin a folder? Could you please give me an example of Doc and DocBin or could you share a link/document where the procedure is explained.

**Reply by ssherlins** [reply]
Hi @walidamamou, is there a forum/community for LayoutLM? If yes, could you please share the link?

**Reply by walidamamou** [reply]
Hi @ssherlins there is a discussion about Layoutlm in [here](https://www.kaggle.com/jpmiller/layoutlm-starter). Otherwise, I am actually working on training a model for invoice extraction, my email is walidamamou87@gmail.com

**Reply by ssherlins** [reply]
Thank you @walidamamou.

**Reply by polm** [reply]
Here's a very simple training data creation script.

```
from spacy.tokens import Doc, DocBin
import spacy

# For this toy example there's just one doc, but you should have several
# hundred at a minimum.
DATA = [
   {"text": "I like cheese.", "entities": [[7, 13, "FOOD"]]}
   ]

nlp = spacy.blank("en")
# A DocBin is basically a special list that will save docs on disk
doc_bin = DocBin(attrs=["ENT_IOB", "ENT_TYPE"])
for eg in DATA:
    # for each example, make a Doc that looks like your desired output
    doc = nlp(eg["text"])
    doc.ents = [
        doc.char_span(s[0], s[1], label=s[2])
        for s in eg.get("entities", [])
    ]
    # add the Doc to the DocBin
    doc_bin.add(doc)
# "train.spacy" is the name of a file which will hold your training data
doc_bin.to_disk("train.spacy")
print(f"Processed {len(doc_bin)} documents")
```

Note that spaCy has no way to make use of image data.

**Reply by polm** [reply]
Oh, one other thing - in your annotations, it looks like you were treating each line in your document as a spaCy Doc, but it would probably make more sense to treat them as one big Doc.

**Reply by ssherlins** [reply]
Yes, that is because after OCR the output that I get has a lot of empty lines in between. I believe appending the lines as a single string should do the job. Also does utf-8 encoding on data affect the training?

**Reply by polm** [reply]
UTF8 is the only supported encoding.
